static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,409 @@
+			// occasional fastdom bomb in the middle.
+			listener.reduce<Promise<unknown>>(
+				(func, listener) =>
+					func.then((ret) => {
+						const thisRet = listener(
+							data.value,
+							ret,
+							getIframe(data),
+						);
+						return thisRet === undefined ? ret : thisRet;
+					}),
+				Promise.resolve(),
+			);
+
+		return promise
+			.then((response) => {
+				respond(data.id, event.source, null, response);
+			})
+			.catch((ex) => {
+				reportError(ex, {
+					feature: 'native-ads',
+				});
+				respond(data.id, event.source, formatError(error500, ex), null);
+			});
+	} else if (typeof listener === 'function') {
+		// We found a persistent listener, to which we just delegate
+		// responsibility to write something. Anything. Really.
+		listener(data.value, respond, getIframe(data));
+	} else {
+		// If there is no routine attached to this event type, we just answer
+		// with an error code
+		respond(data.id, event.source, formatError(error405, data.type), null);
+	}
+};
+
+const on = (window: WindowProxy) => {
+	window.addEventListener('message', (event) => void onMessage(event));
+};
+
+const off = (window: WindowProxy) => {
+	window.removeEventListener('message', (event) => void onMessage(event));
+};
+
+/**
+ * Register a callback for a given type of iframe message
+ *
+ * @param type The `type` of message to register against
+ * @param callback The callback to register that will receive messages of the given type
+ * @param options Options for the target window and whether the callback is persistent
+ */
+export const register: RegisterListener = (type, callback, options): void => {
+	if (REGISTERED_LISTENERS === 0) {
+		on(options?.window ?? window);
+	}
+
+	// Persistent LISTENERS are exclusive
+	if (options?.persist) {
+		LISTENERS[type] = callback;
+		REGISTERED_LISTENERS += 1;
+	} else {
+		const listeners = LISTENERS[type] ?? [];
+
+		if (Array.isArray(listeners) && !listeners.includes(callback)) {
+			LISTENERS[type] = [...listeners, callback];
+			REGISTERED_LISTENERS += 1;
+		}
+	}
+};
+
+/**
+ * Unregister a callback for a given type
+ *
+ * @param type The type of message to unregister against. An iframe will send
+ * messages annotated with the type
+ * @param callback Optionally include the original callback. If this is included
+ * for a persistent callback this function will be unregistered. If it's
+ * included for a non-persistent callback only the matching callback is removed,
+ * otherwise all callbacks for that type will be unregistered
+ * @param options Option for the target window
+ */
+export const unregister: UnregisterListener = (type, callback, options) => {
+	const listeners = LISTENERS[type];
+
+	if (listeners === undefined) {
+		throw new Error(formatError(error405, type).message);
+	} else if (listeners === callback) {
+		LISTENERS[type] = undefined;
+		REGISTERED_LISTENERS -= 1;
+	} else if (Array.isArray(listeners)) {
+		if (callback === undefined) {
+			LISTENERS[type] = [];
```

Comment:
To remove `LISTENERS[type]` we do a mix of setting it to be `undefined` or `[]`. Maybe it should use `delete` everywhere?

To my eyes this particular function is particularly hard to read. 

Perhaps it could be improved it by:
- breaking up the function into smaller ones
- a function which takes a filter function and then removes those matches
- more granular comments

Reply:
Sorry I realised that this is a TS conversion ticket so the above comments can be pushed to a "make it better" ticket.

Reply:
No worries! It's a useful insight. I'm going to collect up a list of follow-ups in a separate comment that we can address separate to this PR.